addons/forum-post-save/userscript.js

```diff
@@ -0,0 +1,59 @@
+export default ({ addon }) => {
+  const topicId = location.pathname.split("/")[3];
+  const box = document.querySelector(".markItUpContainer .markItUpEditor");
+
+  // Purge cache which is over two weeks old
+  const cache = _getAllCache();
+  const now = Date.now();
+
+  let madeAnyChanges = false;
+  for (const topicId of Object.keys(cache)) {
+    if (now - cache[topicId].stamp > 1209600000) {
+      // two weeks (1000 * 60 * 60 * 24 * 14): (ms/s) * (s/m) * (m/h) * (h/d) * (d/w)
+      delete cache[topicId];
+      madeAnyChanges = true;
+    }
+  }
+  if (madeAnyChanges) {
+    localStorage.setItem("sa-forum-post-save", JSON.stringify(cache)); // dont use updateCache here since it won't remove deleted items
+  }
+  if (typeof cache[topicId]?.cache === "string") {
+    box.value = cache[topicId].cache;
+  }
+
+  let lastSaved = 0;
+  box.addEventListener("input", (e) => {
+    if (!addon.self.disabled && Date.now() - lastSaved >= 250) {
```

Comment:
```suggestion
    if (!addon.self.disabled && Date.now() - lastSaved >= 5000) {
```
As Samq64 said, I don't think that we really need to save it four times per second. Even saving every five seconds, there's not much to lose if you close the page.

The only problem with that is that, with this if-then condition, the last part of someone's message won't get saved if they wrote it in the middle of the cooldown time. Imagine the following scenario:
- User writes part of their post
- Draft saves
- User spends another fraction of a second writing
- (Notice that the last part of the post doesn't save because during that last part of writing, the cooldown time wasn't over, therefore if-then condition was false, meaning it didn't trigger the save)

<sub>(Hope that makes sense.)</sub>
Might need to do some sort of "wait until" thing. (Also, don't expect to replicate the above because... 👇)

Also, even with this code, it's somehow actually saving _every single keystroke_ as if the cooldown wasn't even there (I added a `console.log` action to test it), when it's expected to only save if it's been 250 ms, as demonstrated by the code, so it seems like the cooldown doesn't work for some reason. That issue is something to work on too.

Reply:
Apparently `lastSaved` is initialized once (to 0) and then never updated. That causes the second problem.

Reply:
I don't think it should even check for keystrokes, just keep looping and add a timeout to the function.